Fix phpstan missingType.generics

```
 ------ ------------------------------------------------------------------------------------------------------------------ 
  Line   src/DBmysqlIterator.php                                                                                           
 ------ ------------------------------------------------------------------------------------------------------------------ 
  46     Class DBmysqlIterator implements generic interface SeekableIterator but does not specify its types: TKey, TValue  
         🪪  missingType.generics                                                                                          
 ------ ------------------------------------------------------------------------------------------------------------------ 

 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Glpi/Asset/Asset.php                                                                                                                                             
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  119    Method Glpi\Asset\Asset::getDefinitionClassInstance() return type with generic class Glpi\CustomObject\AbstractDefinition does not specify its types: ConcreteClass  
         🪪  missingType.generics                                                                                                                                             
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   src/Glpi/Asset/AssetDefinition.php                                                                                                                                                        
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  86     Method Glpi\Asset\AssetDefinition::getDefinitionManagerClass() return type with generic class Glpi\CustomObject\AbstractDefinitionManager does not specify its types: ConcreteDefinition  
         🪪  missingType.generics                                                                                                                                                                  
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Glpi/CustomObject/AbstractDefinition.php                                                                                                                                                        
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  99     Method Glpi\CustomObject\AbstractDefinition::getDefinitionManagerClass() return type with generic class Glpi\CustomObject\AbstractDefinitionManager does not specify its types: ConcreteDefinition  
         🪪  missingType.generics                                                                                                                                                                            
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Glpi/CustomObject/AbstractDefinitionManager.php                                                                                                                                                       
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  51     Method Glpi\CustomObject\AbstractDefinitionManager::getInstance() return type with generic class Glpi\CustomObject\AbstractDefinitionManager does not specify its types: ConcreteDefinition               
         🪪  missingType.generics                                                                                                                                                                                  
  112    Method Glpi\CustomObject\AbstractDefinitionManager::registerDefinition() has parameter $definition with generic class Glpi\CustomObject\AbstractDefinition but does not specify its types: ConcreteClass  
         🪪  missingType.generics                                                                                                                                                                                  
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Glpi/Form/ServiceCatalog/HomeSearchManager.php                                                                                                                                                                    
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  50     Property Glpi\Form\ServiceCatalog\HomeSearchManager::$providers with generic interface Glpi\Form\ServiceCatalog\Provider\LeafProviderInterface does not specify its types: T                                          
         🪪  missingType.generics                                                                                                                                                                                              
  85     Method Glpi\Form\ServiceCatalog\HomeSearchManager::registerPluginProvider() has parameter $provider with generic interface Glpi\Form\ServiceCatalog\Provider\LeafProviderInterface but does not specify its types: T  
         🪪  missingType.generics                                                                                                                                                                                              
  93     Method Glpi\Form\ServiceCatalog\HomeSearchManager::getProviders() return type with generic interface Glpi\Form\ServiceCatalog\Provider\LeafProviderInterface does not specify its types: T                            
         🪪  missingType.generics                                                                                                                                                                                              
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Glpi/Form/ServiceCatalog/Provider/LeafProviderInterface.php                                                                                                                        
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  43     Interface Glpi\Form\ServiceCatalog\Provider\LeafProviderInterface extends generic interface Glpi\Form\ServiceCatalog\Provider\ItemProviderInterface but does not specify its types: T  
         🪪  missingType.generics                                                                                                                                                               
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   src/Glpi/Form/ServiceCatalog/ServiceCatalogManager.php                                                                                                                                                                    
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  56     Property Glpi\Form\ServiceCatalog\ServiceCatalogManager::$providers with generic interface Glpi\Form\ServiceCatalog\Provider\ItemProviderInterface does not specify its types: T                                          
         🪪  missingType.generics                                                                                                                                                                                                  
  67     Method Glpi\Form\ServiceCatalog\ServiceCatalogManager::registerPluginProvider() has parameter $provider with generic interface Glpi\Form\ServiceCatalog\Provider\ItemProviderInterface but does not specify its types: T  
         🪪  missingType.generics                                                                                                                                                                                                  
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

                                                                                                                        
 [ERROR] Found 12 errors                                                                                                
                                                                                                                        
```

--- a/src/DBmysqlIterator.php
+++ b/src/DBmysqlIterator.php
@@ -42,6 +42,7 @@ use function Safe\preg_split;
 
 /**
  *  Database iterator class for Mysql
+ * @implements SeekableIterator<int, mixed>
  **/
 class DBmysqlIterator implements SeekableIterator, Countable
 {

--- a/src/Glpi/Asset/Asset.php
+++ b/src/Glpi/Asset/Asset.php
@@ -116,6 +116,9 @@ abstract class Asset extends CommonDBTM implements AssignableItemInterface, Stat
         return $definition;
     }
 
+    /**
+     * @return AssetDefinition
+     */
     public static function getDefinitionClassInstance(): AbstractDefinition
     {
         return new AssetDefinition();

--- a/src/Glpi/Asset/AssetDefinition.php
+++ b/src/Glpi/Asset/AssetDefinition.php
@@ -83,6 +83,9 @@ final class AssetDefinition extends AbstractDefinition
         return 'Asset';
     }
 
+    /**
+     * @return class-string<AssetDefinitionManager>
+     */
     public static function getDefinitionManagerClass(): string
     {
         return AssetDefinitionManager::class;

--- a/src/Glpi/CustomObject/AbstractDefinition.php
+++ b/src/Glpi/CustomObject/AbstractDefinition.php
@@ -94,7 +94,7 @@ abstract class AbstractDefinition extends CommonDBTM
 
     /**
      * Get the class name for the definition manager of this type.
-     * @return class-string<AbstractDefinitionManager>
+     * @return class-string<AbstractDefinitionManager<static>>
      */
     abstract public static function getDefinitionManagerClass(): string;
 

--- a/src/Glpi/CustomObject/AbstractDefinitionManager.php
+++ b/src/Glpi/CustomObject/AbstractDefinitionManager.php
@@ -48,6 +48,9 @@ abstract class AbstractDefinitionManager
      */
     private array $definitions = [];
 
+    /**
+     * @phpstan-return static<ConcreteDefinition>
+     */
     abstract public static function getInstance(): self;
 
     /**
@@ -108,6 +111,7 @@ abstract class AbstractDefinitionManager
 
     /**
      * Register a definition.
+     * @phpstan-param ConcreteDefinition $definition
      */
     public function registerDefinition(AbstractDefinition $definition): void
     {

--- a/src/Glpi/Form/ServiceCatalog/HomeSearchManager.php
+++ b/src/Glpi/Form/ServiceCatalog/HomeSearchManager.php
@@ -46,7 +46,7 @@ final class HomeSearchManager
     /** @var int */
     private const MAX_ITEMS_PER_TYPE = 20;
 
-    /** @var LeafProviderInterface[] */
+    /** @var LeafProviderInterface<covariant ServiceCatalogLeafInterface>[] */
     private array $providers;
 
     private bool $providers_are_sorted = false;
@@ -82,6 +82,10 @@ final class HomeSearchManager
         return $items_by_label;
     }
 
+    /**
+     * @param LeafProviderInterface<covariant ServiceCatalogLeafInterface> $provider
+     * @return void
+     */
     public function registerPluginProvider(
         LeafProviderInterface $provider
     ): void {
@@ -89,7 +93,7 @@ final class HomeSearchManager
         $this->providers_are_sorted = false;
     }
 
-    /** @return LeafProviderInterface[] */
+    /** @return LeafProviderInterface<covariant ServiceCatalogLeafInterface>[] */
     private function getProviders(): array
     {
         if (!$this->providers_are_sorted) {

--- a/src/Glpi/Form/ServiceCatalog/Provider/LeafProviderInterface.php
+++ b/src/Glpi/Form/ServiceCatalog/Provider/LeafProviderInterface.php
@@ -39,6 +39,7 @@ use Glpi\Form\ServiceCatalog\ItemRequest;
 
 /**
  * @template T of \Glpi\Form\ServiceCatalog\ServiceCatalogLeafInterface
+ * @extends ItemProviderInterface<T>
  */
 interface LeafProviderInterface extends ItemProviderInterface
 {

--- a/src/Glpi/Form/ServiceCatalog/ServiceCatalogManager.php
+++ b/src/Glpi/Form/ServiceCatalog/ServiceCatalogManager.php
@@ -52,7 +52,7 @@ final class ServiceCatalogManager
     /** @var int */
     public const ITEMS_PER_PAGE = 12;
 
-    /** @var ItemProviderInterface[] */
+    /** @var ItemProviderInterface<covariant ServiceCatalogItemInterface>[] */
     private array $providers;
 
     public function __construct()
@@ -64,6 +64,10 @@ final class ServiceCatalogManager
         ];
     }
 
+    /**
+     * @param ItemProviderInterface<covariant ServiceCatalogItemInterface> $provider
+     * @return void
+     */
     public function registerPluginProvider(
         ItemProviderInterface $provider
     ): void {

